### PR TITLE
[WORKFLOWS-15/34] Create README with onboarding docs

### PR DESCRIPTION
**Pull request by BrunoGrandePhD**
Here's a first draft. Does anyone have feedback? Are there any other sections that I should add? 

The formatted README can be read [**here**](https://github.com/Sage-Bionetworks-Workflows/aws-workflows-nextflow-infra/blob/bgrande/workflows-15/update-readme/README.md).

### CODE REVIEW

**Comment by thomasyu888** [comment]
Thanks! My first look at this is it looks good!
* Is there a requirement that things must go into `develop` first prior to `prod`?
* Is nf-tower only built on `prod`?

File: README.md
Comment (thomasyu888) [inline README.md:17]: resource names prefixed with `dev-`?

Comment (BrunoGrandePhD) [reply README.md:17]: @thomasyu888: Yeah, I made the suggestion [here](https://sagebionetworks.slack.com/archives/CLH1FLNK1/p1627078948024700), and Tess implemented the role for JumpCloud [here](https://github.com/Sage-Bionetworks-IT/organizations-infra/blob/master/org-formation/700-aws-sso/_tasks.yaml#L437-L471). The idea is to grant limited access to IAM such that resources can be created, but important ones cannot be messed up (like the CI service user/role).

**Comment by BrunoGrandePhD** [comment]
> * Is there a requirement that things must go into `develop` first prior to `prod`?

@thomasyu888: I hadn't thought that far ahead yet since we're still setting things up. That said, we could make it a requirement in `CONTRIBUTING.md`. Once NF Tower is up and running in `prod`, we probably want to be conservative with changes to ensure it doesn't go down, similar to the Service Catalog model. 

@tthyer: What do you think? 

> * Is nf-tower only built on `prod`?

@thomasyu888: I expect there to be an instance of NF Tower also running in `develop` for the reason detailed above.

File: README.md
Comment (tthyer) [reply README.md:17]: @thomasyu888 @BrunoGrandePhD Open to other solutions for that. Instead of using that naming scheme we could also create a protected resource policy.

**Comment by tthyer** [comment]
@BrunoGrandePhD @thomasyu888 Regarding things going into dev before prod, the way I have been thinking of this is that if you're developing a new template, it should go into dev first, and generally we may even want sample deployments in dev, such as `sample-nextflow-s3-bucket`. If you're just creating a new sceptre configuration, it can go straight into prod, such as `bgrande-nextflow-s3-bucket`.

The license allows us to have a development and production version of the tower. I would like to see the development deployment really used for development / changes, so we have a place where we're not disturbing users to make changes and updates. PHI or not, use of tower to run workflows can happen in the production account.

File: README.md
Comment (thomasyu888) [reply README.md:17]: I am ok with the current solution, just wanted some clarification.

Comment (BrunoGrandePhD) [reply README.md:17]: I updated the wording.

File: README.md
Comment (thomasyu888) [inline README.md:11]: This is under the assumption people are working on mac or windows right?

File: README.md
Comment (thomasyu888) [inline README.md:13]: I think we may hit the same issue as we had for the provisioner a couple of years ago.  Unfortunately not everyone is proficient with git, so we don't necessarily want that to stop people from spinning up tower project.

File: README.md
Comment (thomasyu888) [inline README.md:67]: Based on this, is the only exception when people create projects in nextflow, since the configurations go straight into prod?

File: README.md
Comment (thomasyu888) [inline README.md:59]: Maybe a final step to try an easy nextflow tower command?

File: README.md
Comment (BrunoGrandePhD) [reply README.md:11]: I updated the Docker link to also include installation instructions for Linux.

File: README.md
Comment (BrunoGrandePhD) [reply README.md:13]: We'll keep an eye on how often Git presents a barrier. My expectation is that the average Tower user will generally be more familiar with Git than the average Service Catalog user. I also don't expect new project to be deployed very often, so admins like myself can readily help out. I added a note clarifying what to do if they don't know how to create a PR.

File: README.md
Comment (BrunoGrandePhD) [reply README.md:67]: I wouldn't say it's the only exception, but it's the main user-facing one. I added this note:

> For example, new instances of the `tower-project.yaml` template don't need to be deployed in `nextflow-dev` first.

File: README.md
Comment (BrunoGrandePhD) [reply README.md:59]: I agree that these docs end abruptly. I'll be extended them once I work through these steps for a real project (_e.g._ imCORE). I'll update the setup script to launch a run using [nextflow-io/hello](https://github.com/nextflow-io/hello) and provide a URL for them to visit to see it complete.

File: README.md
Comment (zaro0508) [inline README.md:15]: hehe, this workflow kinda reminds me of another cool repo i know 😄 .   Unfortunately it was deemed too complicated for scientists 😞 .  I hope this doesn't befall the same fate.  Good thing you guys are willing to support git/github learnings.

File: README.md
Comment (zaro0508) [inline README.md:79]: as a human, how do i get access to these roles?

File: README.md
Comment (zaro0508) [inline README.md:102]: this might be assumed but it's probably important to mention that the CI must never be in a failing state?  otherwise no changes can be updated to the accounts.

File: README.md
Comment (tthyer) [reply README.md:15]: Yes. :grin: We can actually automate this more, but I do think that the subset of users who will be using this service will be more comfortable with doing PRs. We'll see.

File: README.md
Comment (tthyer) [reply README.md:79]: +1 I recommend in onboarding above you have a `0.` about filing the JIRA to ask IT for access to the Tower User role. I wouldn't even mention the other roles in this README -- we could talk about the Developer role in the CONTRIBUTING doc.

File: README.md
Comment (tthyer) [inline README.md:87]: I wonder if we should move this all to the library repo now

**Review by tthyer — CHANGES_REQUESTED** [review]

File: README.md
Comment (BrunoGrandePhD) [reply README.md:56]: I added a step to check for this. Thanks!

File: README.md
Comment (BrunoGrandePhD) [reply README.md:15]: Yeah, @thomasyu888 made the same comment [here](https://github.com/Sage-Bionetworks-Workflows/aws-workflows-nextflow-infra/pull/46#discussion_r700665391). Here was my response:

> We'll keep an eye on how often Git presents a barrier. My expectation is that the average Tower user will generally be more familiar with Git than the average Service Catalog user. I also don't expect new project to be deployed very often, so admins like myself can readily help out. I added a note clarifying what to do if they don't know how to create a PR.

File: README.md
Comment (BrunoGrandePhD) [reply README.md:87]: I created a [JIRA ticket](https://sagebionetworks.jira.com/browse/WORKFLOWS-50) to track this task.

File: README.md
Comment (BrunoGrandePhD) [reply README.md:71]: Since we're gearing the README towards Tower users, I migrated everything but the onboarding instructions to the CONTRIBUTING.md file, with a notice at the top of the README directing prospective contributors.

File: README.md
Comment (thomasyu888) [inline README.md:37]: What does it mean to have read/write access vs read?  Is it that people with read permissions are able to view the workflow run results, but not able to start workflow runs?

Comment (BrunoGrandePhD) [reply README.md:37]: So far, read/write vs read-only only determines what kind of access someone has to the encrypted project bucket. I added a note to clarify. 

That said, Tess and I are working on automating the setup in Tower, and this will likely include the bootstrapping of a Tower workspace within a Sage organization and we'll map these levels of access (read/write vs read-only) to roles in Tower (see screenshot below for available roles). 

image

File: CONTRIBUTING.md
Comment (tthyer) [inline CONTRIBUTING.md:76]: there are now  environment-specific config.yaml files

File: CONTRIBUTING.md
Comment (tthyer) [inline CONTRIBUTING.md:89]: don't forget the pre-commit dependencies

File: CONTRIBUTING.md
Comment (tthyer) [inline CONTRIBUTING.md:92]: remember to add the tower user secrets stored at the repo level too

File: README.md
Comment (tthyer) [inline README.md:84]: Above needs to be revised for #62

**Review by tthyer — CHANGES_REQUESTED** [review]

**Comment by BrunoGrandePhD** [comment]
I'm going to work on this once we merge other PRs that change the onboarding experience. 

Depends on #62 (and a future PR where we switch from hook to CI step)

**Review by tthyer — APPROVED** [review]

**Pull request merged**
